service/history/api/isworkflowtaskvalid/api.go

```
@@ -0,0 +1,88 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package isworkflowtaskvalid
+
+import (
+	"context"
+
+	"go.temporal.io/server/api/historyservice/v1"
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/definition"
+	"go.temporal.io/server/service/history/api"
+	"go.temporal.io/server/service/history/consts"
+	"go.temporal.io/server/service/history/shard"
+)
+
+func Invoke(
+	ctx context.Context,
+	req *historyservice.IsWorkflowTaskValidRequest,
+	shard shard.Context,
+	workflowConsistencyChecker api.WorkflowConsistencyChecker,
+) (resp *historyservice.IsWorkflowTaskValidResponse, retError error) {
+	isValid := false
+	err := api.GetAndUpdateWorkflowWithNew(
+		ctx,
+		req.Clock,
+		api.BypassMutableStateConsistencyPredicate,
+		definition.NewWorkflowKey(
+			req.NamespaceId,
+			req.Execution.WorkflowId,
+			req.Execution.RunId,
+		),
+		func(workflowContext api.WorkflowContext) (*api.UpdateWorkflowAction, error) {
+			isTaskValid, err := isWorkflowTaskValid(workflowContext, req.ScheduledEventId)
+			if err != nil {
+				return nil, err
+			}
+			isValid = isTaskValid
+			return &api.UpdateWorkflowAction{
+				Noop:               true,
+				CreateWorkflowTask: false,
+			}, nil
+		},
+		nil,
+		shard,
+		workflowConsistencyChecker,
+	)
+	return &historyservice.IsWorkflowTaskValidResponse{
+		IsValid: isValid,
+	}, err
+}
+
+func isWorkflowTaskValid(
+	workflowContext api.WorkflowContext,
+	scheduledEventID int64,
+) (bool, error) {
+	mutableState := workflowContext.GetMutableState()
+	if !mutableState.IsWorkflowExecutionRunning() {
+		return false, consts.ErrWorkflowCompleted
```

Can it just return invalid? Why do we need to return an error?

---

we can return not found error or invalid, there is no difference
not found can also be returned if target workflow is gone